File: src/lib/money.ts
// Money is integer cents everywhere. Never floats — 0.1 + 0.2 problems in a
// marketplace ledger are not worth the convenience.

export const DEFAULT_PLATFORM_FEE_PERCENT = Number(process.env.PLATFORM_FEE_PERCENT ?? 10);

export function formatCents(cents: number): string {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: cents % 100 === 0 ? 0 : 2,
    maximumFractionDigits: 2,
  }).format(cents / 100);
}

export function formatCentsExact(cents: number): string {
  return new Intl.NumberFormat("en-US", { style: "currency", currency: "USD" }).format(cents / 100);
}

export function dollarsToCents(dollars: number | string): number {
  const value = typeof dollars === "string" ? Number.parseFloat(dollars) : dollars;
  if (!Number.isFinite(value)) return 0;
  return Math.round(value * 100);
}

export function centsToDollars(cents: number): number {
  return Math.round(cents) / 100;
}

export type FeeSplit = {
  /** What the customer is charged. */
  totalCents: number;
  /** What the platform keeps. */
  platformFeeCents: number;
  /** What lands in the provider's Stripe balance. */
  providerPayoutCents: number;
};

/**
 * $40 service at a 10% marketplace fee => provider $36, platform $4.
 * The fee is floored so rounding never charges the provider an extra cent.
 */
export function splitFee(totalCents: number, feePercent: number): FeeSplit {
  const safeTotal = Math.max(0, Math.round(totalCents));
  const safePercent = Math.min(100, Math.max(0, feePercent));
  const platformFeeCents = Math.floor((safeTotal * safePercent) / 100);
  return {
    totalCents: safeTotal,
    platformFeeCents,
    providerPayoutCents: safeTotal - platformFeeCents,
  };
}

export type DiscountInput = {
  discountType: string;
  discountValue: number;
};

export function applyDiscount(priceCents: number, promo: DiscountInput | null): number {
  if (!promo) return priceCents;
  if (promo.discountType === "PERCENT") {
    return Math.max(0, priceCents - Math.floor((priceCents * promo.discountValue) / 100));
  }
  return Math.max(0, priceCents - promo.discountValue);
}

File: api/src/main/java/app/campusconnect/service/Money.java
package app.campusconnect.service;

import app.campusconnect.domain.DiscountType;
import app.campusconnect.domain.Promotion;

/**
 * All money maths. Integer cents only — a marketplace ledger is the last place
 * you want floating point rounding.
 */
public final class Money {

    private Money() {
    }

    /** What the customer pays, what the platform keeps, what the provider gets. */
    public record FeeSplit(int totalCents, int platformFeeCents, int providerPayoutCents) {
    }

    /**
     * $40 at a 10% fee: provider $36, platform $4.
     * The fee floors, so rounding never quietly costs the provider a cent.
     */
    public static FeeSplit split(int totalCents, double feePercent) {
        int safeTotal = Math.max(0, totalCents);
        double safePercent = Math.min(100, Math.max(0, feePercent));
        int fee = (int) Math.floor(safeTotal * safePercent / 100.0);
        return new FeeSplit(safeTotal, fee, safeTotal - fee);
    }

    public static int applyDiscount(int priceCents, Promotion promotion) {
        if (promotion == null) {
            return priceCents;
        }
        if (promotion.getDiscountType() == DiscountType.PERCENT) {
            int off = (int) Math.floor(priceCents * promotion.getDiscountValue() / 100.0);
            return Math.max(0, priceCents - off);
        }
        return Math.max(0, priceCents - promotion.getDiscountValue());
    }

    public static String format(int cents) {
        return String.format("$%,.2f", cents / 100.0);
    }
}
